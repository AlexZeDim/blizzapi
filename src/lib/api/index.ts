import * as BattleNetApi from './BattleNetApi';
export default BattleNetApi;
