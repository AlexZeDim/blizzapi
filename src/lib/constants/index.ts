import regions from './regions';
import locales from './locales';
import defaultLocales from './defaultLocales';
// export * from './servers';
// export * from './regions';

// TODO: complete this

// interface CONSTANT<Value> {
//   [regionId: string]: Value;
// }

export default class CONSTANTS {
  public static readonly REGIONS = regions;
  public static readonly LOCALES = locales;
  public static readonly DEFAULT_LOCALES = defaultLocales;
  // public static readonly SERVERS = servers;
}


// interface CONSTANT<Value> {
//   [regionId: string]: Value;
// }

// export default class CONSTANTS {
//   public static readonly REGIONS: CONSTANT<string | string[]> = {
//     1: 'us',
//     2: 'eu',
//     3: ['kr', 'tw'],
//     5: 'cn',
//   };

//   public static readonly LOCALES: CONSTANT<string[]> = {
//     1: ['en_US', 'es_MX', 'pt_BR'],
//     2: ['en_GB', 'es_ES', 'fr_FR', 'ru_RU', 'de_DE', 'pt_PT', 'it_IT'],
//     3: ['ko_KR', 'zh_TW'],
//     5: ['zh_CN'],
//   };

//   public static readonly DEFAULT_LOCALES: CONSTANT<number> = {
//     1: 0,
//     2: 0,
//     3: 0,
//     5: 0,
//   };

//   public static readonly SERVERS: CONSTANT<string> = {
//     1: 'https://us.api.blizzard.com',
//     2: 'https://eu.api.blizzard.com',
//     3: 'https://kr.api.blizzard.com',
//     5: 'https://api.blizzard.com.cn',
//   };

//   public static readonly SC2_REALMS: CONSTANT<number[]> = {
//     1: [1, 2],
//     2: [1, 2],
//     3: [1, 2],
//     5: [1],
//   };

//   public static readonly OAUTH_AUTHORIZE_SERVERS: CONSTANT<string> = {
//     1: 'https://us.battle.net/oauth/authorize',
//     2: 'https://eu.battle.net/oauth/authorize',
//     3: 'https://apac.battle.net/oauth/authorize',
//     5: 'https://www.battlenet.com.cn/oauth/authorize',
//   };

//   public static readonly OAUTH_TOKEN_URIS: CONSTANT<string> = {
//     1: 'https://us.battle.net/oauth/token',
//     2: 'https://eu.battle.net/oauth/token',
//     3: 'https://apac.battle.net/oauth/token',
//     5: 'https://www.battlenet.com.cn/oauth/token',
//   };

//   public static readonly OAUTH_CHECK_TOKEN_URIS: CONSTANT<string> = {
//     1: 'https://us.battle.net/oauth/check_token?token=',
//     2: 'https://eu.battle.net/oauth/check_token?token=',
//     3: 'https://apac.battle.net/oauth/check_token?token=',
//     5: 'https://www.battlenet.com.cn/oauth/check_token?token=',
//   };
// }
