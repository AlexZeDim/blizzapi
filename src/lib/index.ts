export * from './utils';
// // todo import battlenet api class

// export {
//   // todo: bnet api class export
//   BattleNetUtils,
// }