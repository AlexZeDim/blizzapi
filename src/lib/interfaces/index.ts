// Still don't know how to avoid shipping this file in dist build
// https://github.com/Microsoft/TypeScript/issues/10908

export interface IRegionIdProperties<Value> {
  readonly [regionId: string]: Value;
}
