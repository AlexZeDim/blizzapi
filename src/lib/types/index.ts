// Still don't know how to avoid shipping this file in dist build
// https://github.com/Microsoft/TypeScript/issues/10908

/** Common */
export type Uri = string;
export type HostUri = string;

/** Regions */
export type RegionId = number;
export type RegionIdKey = string;
export type RegionIdAsNumberOrString = RegionId | RegionIdKey;
export type RegionIdArray = ReadonlyArray<RegionId>;
export type RegionName = string;
export type RegionNameArray = ReadonlyArray<RegionName>;

/** Locales */
export type Locale = string;
export type LocaleArray = ReadonlyArray<Locale>;

/** Default locales */
export type DefaultLocaleIndex = number;
export type DefaultLocaleIndexArray = ReadonlyArray<DefaultLocaleIndex>;
export type RegionApiHostUri = Uri;

/** StarCraft II realms  */
export type Sc2Realm = number;
export type Sc2RealmList = ReadonlyArray<Sc2Realm>;

/** OAuth */
export type AuthorizeUri = Uri;
export type TokenUri = Uri;
export type CheckTokenUri = Uri;
