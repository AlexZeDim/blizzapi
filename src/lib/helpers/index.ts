import 'cross-fetch/polyfill';
import { Uri } from '../types';

/**
 * Validates uri via matching it against regex pattern
 * @function
 * @param {string} uri Fetch request uri starting with http(s) (e.g. https://example.com/api/endpoint)
 * @returns {Boolean} True for valid uri, false for incorrect one
 */
export function validateUri(uri:Uri) {
  // via https://www.regextester.com/94502
  const uriTestRegex = /^(?:http(s)?:\/\/)[\w.-]+(?:\.[\w\.-]+)+[\w\-\._~:/?#[\]@!\$&'\(\)\*\+,;=.]+$/gm;
  uriTestRegex.lastIndex = 0;
  return uriTestRegex.test(uri);
}

/**
 * Performs basic fetch request from given uri
 * @function
 * @param {string} uri Fetch request uri
 * @returns {object} Data returned by requested uri
 */
export function fetchFromUri(uri:Uri) {
  const isUriValid = validateUri(uri);

  if (!isUriValid) {
    throw new RangeError(`'${uri}' is not a valid parameter for fetchFromUri()`);
  }

  return fetch(uri);
}
