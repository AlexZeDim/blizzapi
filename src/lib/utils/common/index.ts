import { ConstantKey, RegionIdOrName, RegionIdAsNumberOrString, RegionName, Uri } from '../../types';
import constants from '../../constants';
import { validateRegionId, validateRegionName, getRegionIdByName } from '../localization/regions';

export const getUriByRegion = (regionIdOrName: RegionIdOrName, constantKey: ConstantKey) => {
  const validRegionId = validateRegionId(regionIdOrName);

  /* tslint:disable no-if-statement */
  if (validRegionId) {
    return getUriByRegionId(regionIdOrName, constantKey);
  }
  /* tslint:enable no-if-statement */

  return getUriByRegionName(regionIdOrName.toString(), constantKey);
};

export const getUriByRegionId = (regionId: RegionIdAsNumberOrString, constantKey: ConstantKey) => {
  const regionIdAsString = regionId.toString();
  const isRegionIdValid = validateRegionId(regionIdAsString);

  /* tslint:disable no-if-statement */
  if (!isRegionIdValid) {
    throw new RangeError(
      `${regionIdAsString} is not a valid parameter for getUriByRegionId(${regionId}, '${constantKey}')`,
    );
  }
  /* tslint:enable no-if-statement */

  return <Uri>constants[constantKey][regionIdAsString];
};

/**
 * Returns uri for given region name from specified constant object
 *
 * @param regionName Region name as two-letter string (e.g. 'us')
 * @param constantKey Constant key
 * @return Uri as a string
 */
export const getUriByRegionName = (regionName: RegionName, constantKey: ConstantKey) => {
  const isRegionNameValid = validateRegionName(regionName);

  /* tslint:disable no-if-statement */
  if (!isRegionNameValid) {
    throw new RangeError(
      `${regionName} is not a valid parameter for getUriByRegionName(${regionName}, '${constantKey}')`,
    );
  }
  /* tslint:enable no-if-statement */

  const regionId = getRegionIdByName(regionName);

  return <Uri>constants[constantKey][regionId];
};
