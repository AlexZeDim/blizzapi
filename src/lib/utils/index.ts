import * as localization from './localization';

import { RegionName, RegionIdAsNumberOrString, Locale } from '../types';

/** Regions */

export function getAllRegions() {
  return localization.getAllRegions();
}

export function getAllRegionNames() {
  return localization.getAllRegionNames();
}

export function getAllRegionIds() {
  return localization.getAllRegionIds();
}

export function getRegionNameById(regionId: RegionIdAsNumberOrString) {
  return localization.getRegionNameById(regionId);
}

export function validateRegionId(regionId: RegionIdAsNumberOrString) {
  return localization.validateRegionId(regionId);
}

export function getRegionIdByName(regionName: RegionName) {
  return localization.getRegionIdByName(regionName);
}

export function validateRegionName(regionName: RegionName) {
  return localization.validateRegionName(regionName);
}

/** Locales */

export function getAllLocales() {
  return localization.getAllLocales();
}

export function getAllLocaleNames() {
  return localization.getAllLocaleNames();
}

export function getLocalesByRegionId(regionId: RegionIdAsNumberOrString) {
  return localization.getLocalesByRegionId(regionId);
}

export function checkIfLocaleLooksValid(locale: Locale) {
  return localization.checkIfLocaleLooksValid(locale);
}

export function validateLocale(locale: Locale) {
  return localization.validateLocale(locale);
}

export function isLocaleValidForRegionId(locale: Locale, regionId: RegionIdAsNumberOrString) {
  return localization.isLocaleValidForRegionId(locale, regionId);
}

/** Default locales */

export function getAllDefaultLocaleIndexes() {
  return localization.getAllDefaultLocaleIndexes();
}

export function getAllDefaultLocaleNames() {
  return localization.getAllDefaultLocaleNames();
}

export function getDefaultLocaleIndexForRegionId(regionId: RegionIdAsNumberOrString) {
  return localization.getDefaultLocaleIndexForRegionId(regionId);
}

export function getDefaultLocaleNameForRegionId(regionId: RegionIdAsNumberOrString) {
  return localization.getDefaultLocaleNameForRegionId(regionId);
}

/** StarCraft II realms */

export function getAllSc2Realms() {
  return localization.getAllSc2Realms();
}
