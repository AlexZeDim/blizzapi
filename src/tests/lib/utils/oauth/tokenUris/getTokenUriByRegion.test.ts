import * as utils from '../../../../../lib/utils';
// import constants from '../../../../../lib/constants';

const { getTokenUriByRegion } = utils;

const regionNames = [
  'us',
  'eu',
  'kr',
  'tw',
  'cn',
  'us',
  'EU',
  'KR',
  'TW',
  'CN',
  'Us',
  'Eu',
  'Kr',
  'Tw',
  'Cn',
  'uS',
  'eU',
  'kR',
  'tW',
  'cN',
];

const wrongRegionNames = [
  'a1',
  'adasf',
  '1abc',
  '!@#$',
  '><1',
];

const regionIds = [
  1,
  2,
  3,
  5,
  '1',
  '2',
  '3',
  '5',
];

const wrongRegionIds = [
  '10',
  '9999',
  '23232',
  'a1',
  'adasf',
  '1abc',
  '!@#$',
  '><1',
  10,
  34,
  6393,
  9999,
];

describe('getTokenUriByRegion()', () => {
  test('should be defined', () => {
    expect(getTokenUriByRegion).toBeDefined();
  });

  test('should be of type "function"', () => {
    expect(typeof getTokenUriByRegion).toBe('function');
  });

  regionIds.forEach(regionId => {
    test('should return correct value for ${regionId} and ${constantKey} as valid parameters', () => {
      expect(getTokenUriByRegion(regionId)).toMatchSnapshot();
    });
  });

  wrongRegionIds.forEach(wrongRegionId => {
      test('should throw RangeError for ${wrongRegionId} as invalid parameter', () => {
        expect(() => getTokenUriByRegion(wrongRegionId)).toThrow(RangeError);
      });
    });

  regionNames.forEach(regionName => {
    test('should return correct value for ${regionName} as valid region name', () => {
      expect(getTokenUriByRegion(regionName)).toMatchSnapshot();
    });
  });

  wrongRegionNames.forEach(wrongRegionName => {
    test('should throw RangeError for invalid region name', () => {
      expect(() => getTokenUriByRegion(wrongRegionName)).toThrow(RangeError);
    });
  });
});
