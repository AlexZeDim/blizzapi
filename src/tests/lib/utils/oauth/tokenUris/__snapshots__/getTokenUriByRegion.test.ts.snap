// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTokenUriByRegion() should return correct value for \${regionId} and \${constantKey} as valid parameters 1`] = `"https://us.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionId} and \${constantKey} as valid parameters 2`] = `"https://eu.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionId} and \${constantKey} as valid parameters 3`] = `"https://apac.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionId} and \${constantKey} as valid parameters 4`] = `"https://www.battlenet.com.cn/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionId} and \${constantKey} as valid parameters 5`] = `"https://us.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionId} and \${constantKey} as valid parameters 6`] = `"https://eu.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionId} and \${constantKey} as valid parameters 7`] = `"https://apac.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionId} and \${constantKey} as valid parameters 8`] = `"https://www.battlenet.com.cn/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 1`] = `"https://us.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 2`] = `"https://eu.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 3`] = `"https://apac.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 4`] = `"https://apac.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 5`] = `"https://www.battlenet.com.cn/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 6`] = `"https://us.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 7`] = `"https://eu.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 8`] = `"https://apac.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 9`] = `"https://apac.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 10`] = `"https://www.battlenet.com.cn/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 11`] = `"https://us.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 12`] = `"https://eu.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 13`] = `"https://apac.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 14`] = `"https://apac.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 15`] = `"https://www.battlenet.com.cn/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 16`] = `"https://us.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 17`] = `"https://eu.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 18`] = `"https://apac.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 19`] = `"https://apac.battle.net/oauth/token"`;

exports[`getTokenUriByRegion() should return correct value for \${regionName} as valid region name 20`] = `"https://www.battlenet.com.cn/oauth/token"`;
